crates/sui-network/src/discovery/mod.rs

```diff
@@ -0,0 +1,448 @@
+
+        match peer_event {
+            Ok(PeerEvent::NewPeer(peer_id)) => {
+                if let Some(peer) = self.network.peer(peer_id) {
+                    self.state
+                        .write()
+                        .unwrap()
+                        .connected_peers
+                        .insert(peer_id, ());
+
+                    // If we haven't figured out what our external address is, ask this peer
+                    if let Some(our_info) = &self.state.read().unwrap().our_info {
+                        if our_info.external_socket_address.is_none() {
+                            self.tasks.spawn(query_our_external_socket_address(
+                                peer.clone(),
+                                self.state.clone(),
+                            ));
+                        }
+                    }
+
+                    // Query the new node for any peers
+                    self.tasks
+                        .spawn(query_peer_for_their_known_peers(peer, self.state.clone()));
+                }
+            }
+            Ok(PeerEvent::LostPeer(peer_id, _)) => {
+                self.state.write().unwrap().connected_peers.remove(&peer_id);
+            }
+
+            Err(RecvError::Closed) => {
+                panic!("PeerEvent channel shouldn't be able to be closed");
+            }
+
+            Err(RecvError::Lagged(_)) => {
+                trace!("State-Sync fell behind processing PeerEvents");
+            }
+        }
+    }
+
+    fn handle_tick(&mut self, _now: std::time::Instant, now_unix: u64) {
+        self.update_our_info_timestamp(now_unix);
+
+        self.tasks
+            .spawn(query_connected_peers_for_their_known_peers(
+                self.network.clone(),
+                self.state.clone(),
+            ));
+
+        // Cull old peers older than a day
+        self.state
+            .write()
+            .unwrap()
+            .known_peers
+            .retain(|_k, v| now_unix.saturating_sub(v.timestamp_ms) < ONE_DAY_MILLISECONDS);
+
+        // Clean out the pending_dials
+        self.pending_dials.retain(|_k, v| !v.is_finished());
+
+        // Spawn some dials
+        let state = self.state.read().unwrap();
+        let mut eligible = state
+            .known_peers
+            .clone()
+            .into_iter()
+            .filter(|(peer_id, info)| {
+                peer_id != &self.network.peer_id() &&
+                (!info.addresses.is_empty() || info.external_socket_address.is_some()) // Peer has addresses we can dial
+                && !state.connected_peers.contains_key(peer_id) // We're not already connected
+                && !self.pending_dials.contains_key(peer_id) // There is no pending dial to this node
+            })
+            .collect::<Vec<_>>();
+
+        // randomize the order
+        let mut rng = <rand::rngs::StdRng as rand::SeedableRng>::from_entropy();
+        rand::seq::SliceRandom::shuffle(eligible.as_mut_slice(), &mut rng);
```

Comment:
instead of shuffling the entire eligible list when you might only need part of it, is it possible to use `choose_multiple`?